src/CFamily/CMake/VsDevCmdEnvironmentProvider.cs

```diff
@@ -0,0 +1,231 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Abstractions;
+using System.Text;
+using Microsoft.VisualStudio;
+using SonarLint.VisualStudio.CFamily.SystemAbstractions;
+using SonarLint.VisualStudio.Infrastructure.VS;
+using SonarLint.VisualStudio.Integration;
+using SonarLint.VisualStudio.Integration.Helpers;
+
+namespace SonarLint.VisualStudio.CFamily.CMake
+{
+    internal interface IVsDevCmdEnvironmentProvider
+    {
+        /// <summary>
+        /// Returns the environment settings set by VsDevCmd.bat
+        /// </summary>
+        /// <param name="scriptParams">Any additional parameters to pass to the batch file. Can be null.</param>
+        IReadOnlyDictionary<string, string> Get(string scriptParams);
+    }
+
+    internal class VsDevCmdEnvironmentProvider : IVsDevCmdEnvironmentProvider
+    {
+        /// <summary>
+        /// Relative path from the VS installation root to the batch file
+        /// </summary>
+        /// <remarks>This path works for VS2015+</remarks>
+        private const string RelativePathToBatchFile = "Common7\\Tools\\VsDevCmd.bat";
+
+        private const int SCRIPT_TIMEOUT_MS = 4000;
+
+        private readonly IVsInfoService vsInfoService;
+        private readonly ILogger logger;
+
+        // Interfaces for testing
+        private readonly IProcessFactory processFactory;
+        private readonly IFileSystem fileSystem;
+
+
+        public VsDevCmdEnvironmentProvider(IVsInfoService vsInfoService, ILogger logger)
+            : this(vsInfoService, logger, new ProcessFactory(), new FileSystem())
+        {
+            this.vsInfoService = vsInfoService;
+            this.logger = logger;
+        }
+
+        internal /* for testing */ VsDevCmdEnvironmentProvider(IVsInfoService vsInfoService, ILogger logger,
+            IProcessFactory processFactory, IFileSystem fileSystem)
+        {
+            this.vsInfoService = vsInfoService;
+            this.logger = logger;
+            this.processFactory = processFactory;
+            this.fileSystem = fileSystem;
+        }
+
+        public IReadOnlyDictionary<string, string> Get(string scriptParams)
+        {
+            try
+            {
+                string filePath = Path.Combine(vsInfoService.InstallRootDir, RelativePathToBatchFile);
+                LogDebug($"VsDevCmd location: {filePath}");
+
+                if (!fileSystem.File.Exists(filePath))
+                {
+                    logger.WriteLine(Resources.VsDevCmd_FileNotFound, filePath);
+                    return null;
```

Comment:
just a question of usage, returning empty vs null.

Reply:
Difficult to say which is more convenient at this point. We can always change it later when we have a real consumer.